Dispersion and reverse dispersion probability estimators

kahaaga opened the pull request:
# What is this PR? 

This PR does three things:
- Introduces `Dispersion` as a probability estimator (fixes #82), removing it as a direct entropy estimator and replacing it with `renyi_entropy(x, Dispersion(); kwargs...)`.
- Introduces the *complexity measure* `reverse_dispersion` (entropy), which is not included as a separate probability estimator, because the probability estimation step is identical to `Dispersion`, and the final quantities on which the reverse dispersion entropy is computed is a combination of distance information and probabilities, so we can't use `renyi_entropy` directly.
- Moves the tests for these two estimators in a separate file, following the convention agreed upon in #89 .

## References
Li, Y., Gao, X., & Wang, L. (2019). Reverse dispersion entropy: a new complexity measure for sensor signal. Sensors, 19(23), 5203.

Datseris commented on src/probabilities_estimators/dispersion/dispersion_reverse.jl:
```diff
@@ -0,0 +1,55 @@
+function distance_to_whitenoise(𝐩::Probabilities, N, m)
+    # We can safely skip non-occurring symbols, because they don't contribute
+    # to the sum in eq. 3 in Li et al. (2019)
+    return sum(𝐩[i]^2 for i in eachindex(𝐩)) - 1/(N^m)
```
`sum(𝐩[i]^2 for i in eachindex(𝐩))` could be `sum(abs2, p)` instead.

Datseris commented on test/dispersion.jl:
```diff
@@ -0,0 +1,50 @@
+        ps = probabilities(x, est)
+        @test ps isa Probabilities
+
+        de = entropy_renyi(x, est, q = 1, base = 2, normalize = false)
+        de_norm = entropy_renyi(x, est, q = 1, base = 2, normalize = true)
+        @test de isa Real
+        @test de_norm isa Real
+        @test de >= 0.0
+        @test de_norm >= 0.0
```
Can we devise any analytically resolved / meanigful tests for this block? We might as well do proper tests in new PRs. 

`de isa Real` is an interface test. It doesn't test the `ReverseDispersion()` method, it tests the `renyi_entropy` interface. Same with `ps isa Probabilities`. We would need tests that actually test whether the algorithm does as it should by devising a scenario where we know what the output should be a priori.

kahaaga replied:
Good points. For the normalized case, a meaningful test is that the resulting entropy value is on in the interval `[0,1]`. 

It is also possible to construct analytical test cases for this method, but I propose we do so in a different PR.

Datseris commented on src/complexity_measures/reverse_dispersion_entropy.jl:
```diff
@@ -0,0 +1,35 @@
+    # we're not dealing with probabilities anymore.
+    Hrde = distance_to_whitenoise(p, s.n_categories, m)
+
+    if normalize
```
So this keyword is not discussed in the docstring. Is it internal?

kahaaga commented:
@Datseris I think this is good to go from my side. We can deal with the test cases in a separate PR (we also have an open issue to improve tests (#85).

Note that I made some changes regarding the reverse dispersion entropy (see initial comment for the PR). It is a *complexity* estimator, not an entropy estimator.

Datseris commented:
If we want to follow good development practices, we should add tests in this PR, not in another one. A new feature is ideally accompanied by both documentation and tests. Is there any time pressure to merge this immediately for some reason? If not, we should keep it open until we can get around and add tests. 

Especially if we want to write a paper about this, a big selling point we can have is rigorous tests, which, at least from what I've seen so far, all competing libraries are completely void of. So the best practice moving forwards is to demand rigorous tests for all new features. From a scientific perspective this is even more important actually.

kahaaga commented:
Sure, we can wait!

This PR is not extremely urgent, but I have an estimator for TransferEntropy.jl waiting that relies on this estimator (which I want in the CausalityTools paper). However, why not take the opportunity to write robust, analytical tests for both? It’ll probably take a few week or so before I get to addressing this, but it’ll be worth the effort.

kahaaga commented:
Hey, @Datseris 

I actually had analytical tests based on the original papers in an uncommited PR that I had completely forgotten about. I've updated this PR with these test cases (note: the example in the Rostaghi & Azami paper is not correct, but the tests here are adjusted to reflect this).

In addition to adding analytical tests, I've also:
- Added an example to `Entropy.jl examples` comparing the dispersion entropy and reverse dispersion entropy for a noisy signal with a single strong impulse.
- Updated the docstring for `GaussianSymbolization`, so that all docs regarding this symbolization procedure doesn't need to be repeated elsewhere.
- Changed the named argument for `GaussianSymbolization` from `n_categories` to `c`, to conform with the notation in the Rostaghi & Azami and Li et al papers (from which this symbolization approach came).
- Outlined briefly the algorithms in the `Dispersion` and `reverse_dispersion` docstrings.
- Clarified confusing notation in the `reverse_dispersion` docstring.
- Added reproducible examples to docstrings, using `jldoctest`. This way, we can test the examples while developing by running `doctest(Entropies)`.

Datseris commented on src/complexity_measures/reverse_dispersion_entropy.jl:
````diff
@@ -0,0 +1,91 @@
+Because ``H_{rde} \\geq 0``, ``H_{rde}`` can therefore be said to be a measure of how far
+the dispersion pattern probability distribution is from white noise.
+
+## Example
+
+```jldoctest reverse_dispersion_example; setup = :(using Entropies)
+julia> x = repeat([0.5, 0.7, 0.1, -1.0, 1.11, 2.22, 4.4, 0.2, 0.2, 0.1], 10);
+
+julia> c, m = 3, 5;
+
+julia> reverse_dispersion(x, s = GaussianSymbolization(c = c), m = m, normalize = false)
+0.11372331532921814
+```
````
I am not sure about this. I mean how much does it help really? It just shows the function call, and that it outputs a number, which is anyways understood by the first sentence of the docstring. Unlike Julia base functions, whose ouput can be deduced, and hence these tiny snippets actually make sure the user gets what's going on, here it's not like a user could deduce the `0.11...`. Furthermore, there is already an example in the docs. 

More importantly, what are the principles that decide which methods have examples in their docstrings and which do not? Are we perhaps confusing the user? One of the principles of good documentation is that information should be where it is expected to be. So, are we creating the expectation that examples will be found in the docpage "Examples", or at the end of the docstrings?

My vote is to stick with one mean of showing examples, and the vote for that is the dedicated examples page.

---

P.s.: The argument of using this as a means of tests is totally unconvincing for me, that's why we write tests in the `test` folder anyways, so if this tests what is already in the test folder, then why test it twice, and if it tests something that isn't in the test folder, well, why wasn't this thing in the test folder in the first place.

Datseris commented on src/entropies/convenience_definitions.jl:
```diff
@@ -53,6 +53,20 @@ function entropy_wavelet(x; wavelet = Wavelets.WT.Daubechies{12}(), base = MathC
 
-function entropy_dispersion(args...)
+"""
+    entropy_dispersion(x; m = 2, τ = 1, s = GaussianSymbolization(3),
+        base = MathConstants.e)
```
In `reverse_dispersion` arguments are positional, here they are keyword. I'd say we stick with keyword.

kahaaga replied:
> In reverse_dispersion arguments are positional, here they are keyword. I'd say we stick with keyword.

Yes, definitely. As I commented above, I unintentionally missed the semi-colon.

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
```diff
@@ -0,0 +1,172 @@
+categories for the Gaussian symbol mapping. See below for an in-depth description of
+dispersion pattern construction and usage.
+
+# Algorithm
```
```suggestion
## Description
```

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
```diff
@@ -0,0 +1,172 @@
+are ``c^m`` possible dispersion patterns. This number is used for normalization when
+computing dispersion entropy.
+
+## Computing dispersion probabilities and entropy
```
```suggestion
### Computing dispersion probabilities and entropy
```

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
```diff
@@ -0,0 +1,172 @@
+See also: [`entropy_dispersion`](@ref), [`GaussianSymbolization`](@ref).
+
+
+!!! note
+    ## Why "dispersion patterns"?
```
```suggestion
!!! note "Why 'dispersion patterns'?"
```

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
```diff
@@ -0,0 +1,172 @@
+[^Rostaghi2016]: Rostaghi, M., & Azami, H. (2016). Dispersion entropy: A measure for time-series analysis. IEEE Signal Processing Letters, 23(5), 610-614.
+[^Li2018]: Li, G., Guan, Q., & Yang, H. (2018). Noise reduction method of underwater acoustic signals based on CEEMDAN, effort-to-compress complexity, refined composite multiscale dispersion entropy and wavelet threshold denoising. Entropy, 21(1), 11.
+"""
+Base.@kwdef struct Dispersion <: ProbabilitiesEstimator
+    s = GaussianSymbolization(c = 5)
+    m = 2
+    τ = 1
+    check_unique = false
+    normalize = true
+end
```
Why is it that none of the fields here is type stable...? I don't think that was intended, right? All fields here should be type stable.

Datseris commented on test/dispersion.jl:
```diff
@@ -0,0 +1,83 @@
+
+    @testset "Dispersion" begin
+
+        @testset "Internals" begin
```
The property **robust** of a good test suite means that one tests that the expected functionality is met, not how it is met. Subsequently, only exported API should be used in the test suite. The internals are not tested, their final outcome is. See also https://github.com/JuliaDynamics/Entropies.jl/issues/85

Datseris commented on test/dispersion.jl:
```diff
@@ -0,0 +1,83 @@
+        # 1 - 1/(n_classes^m). When normalizing to this value, the RDE should be 1.0.
+        m, n_classes = 2, 2
+        single_element_dist = Probabilities([1.0, 0.0, 0.0, 0.0])
+        Hrde_maximal = Entropies.distance_to_whitenoise(single_element_dist, n_classes, m,
+            normalize = false)
+        Hrde_maximal_norm = Entropies.distance_to_whitenoise(single_element_dist, n_classes, m,
+            normalize = true)
```
Here internal functions are also used. Perhaps this function should be documented and exported, and put into the Utils page?

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
```diff
@@ -0,0 +1,172 @@
+Relative frequencies of dispersion patterns are computed using the symbolization scheme
+`s` with embedding dimension `m` and embedding delay `τ`. Recommended parameter
+values[^Li2018] are `m ∈ [2, 3]`, `τ = 1` for the embedding, and `c ∈ [3, 4, …, 8]`
+categories for the Gaussian symbol mapping. See below for an in-depth description of
+dispersion pattern construction and usage.
```
```suggestion
categories for the Gaussian symbol mapping.
```
The description follows immediatelly, so this sentence isn't necessary :)

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
````diff
@@ -0,0 +1,172 @@
+0.6716889280681666
+```
+
+If `normalize == true`, then when used in combination with [`entropy_renyi`](@ref),
+the dispersion entropy is normalized to `[0, 1]`. Normalization is only
+defined when `q == 1`.
````
Hmmmm, here this is something we can discuss extensively. The normalization process. In general, most entropies that have a _known total number of symbols_ can be normalized versus the entropy of uniform probability for each symbol. E.g., for ordinal patterns this is also known a-priory.

Isn't it better for us to devise a general API for normalizing entropies in this regard, instead of doing it individual for each single estimator?

Here is what I propose: define a function `alphabet_length(::ProbabilitiesEstimator)` or `word_length` or whatever information theory people call it. It returns the total number of possible words. Not just the ones that actually exist in the data. If it is not known a-priori, it returns `nothing` instead. Then, we define a function `entropy_normalized` that returns the renyi entropy but normalized (i.e., divided by `log(alphabet_length)`), i.e., normalized versus uniform distribution.

kahaaga replied:
> Isn't it better for us to devise a general API for normalizing entropies in this regard, instead of doing it individual for each single estimator?
> Here is what I propose: define a function alphabet_length(::ProbabilitiesEstimator) or word_length or whatever information theory people call it. It returns the total number of possible words. Not just the ones that actually exist in the data. If it is not known a-priori, it returns nothing instead. Then, we define a function entropy_normalized that returns the renyi entropy but normalized (i.e., divided by log(alphabet_length)), i.e., normalized versus uniform distribution.

I like that idea a lot. I'll implement that in a PR with all my responses to your comments here.

Datseris commented on src/probabilities_estimators/dispersion/dispersion.jl:
````diff
@@ -0,0 +1,172 @@
+``\\sum_i^{c^m} p_i = 1``, can then be estimated by counting and sum-normalising
+the distribution of dispersion patterns among the embedding vectors.
+
+```jldoctest dispersion_example; setup = :(using Entropies)
````
Same concern with the doctests as in the other docstring.

Datseris commented on src/complexity_measures/reverse_dispersion_entropy.jl:
````diff
@@ -0,0 +1,91 @@
+0.11372331532921814
+```
+
+!!! note
+    #### A clarification on notation
+
+    With ambiguous notation, Li et al. claim that
+
+    ``H_{rde} = \\sum_{i = 1}^{c^m} \\left(p_i - \\dfrac{1}{{c^m}} \\right)^2 = \\sum_{i = 1}^{c^m} p_i^2 - \\frac{1}{c^m}.``
+
+    But on the right-hand side of the equality, does the constant term appear within or
+    outside the sum? Using that ``P`` is a probability distribution by
+    construction (in step 4) , we see that the constant must appear *outside* the sum:
+
+    ```math
+    \\begin{aligned}
+    H_{rde} &= \\sum_{i = 1}^{c^m} \\left(p_i - \\dfrac{1}{{c^m}} \\right)^2
+    = \\sum_{i=1}^{c^m} p_i^2 - \\frac{2p_i}{c^m} + \\frac{1}{c^{2m}} \\\\
+    &= \\left( \\sum_{i=1}^{c^m} p_i^2 \\right) - \\left(\\sum_i^{c^m} \\frac{2p_i}{c^m}\\right) + \\left( \\sum_{i=1}^{c^m} \\dfrac{1}{{c^{2m}}} \\right) \\\\
+    &= \\left( \\sum_{i=1}^{c^m} p_i^2 \\right) - \\left(\\frac{2}{c^m} \\sum_{i=1}^{c^m} p_i \\right) +  \\dfrac{c^m}{c^{2m}} \\\\
+    &= \\left( \\sum_{i=1}^{c^m} p_i^2 \\right) - \\frac{2}{c^m} (1) +  \\dfrac{1}{c^{m}} \\\\
+    &= \\left( \\sum_{i=1}^{c^m} p_i^2 \\right) - \\dfrac{1}{c^{m}}.
+    \\end{aligned}
+    ```
````
I'm not sure if there is any point in writing this huge block of text. Why don't we just type the correct formula in the `Description` section and be done with it? Furthermore, do we even _need_ to discuss this at all? I don't think it matters for _the software_ that this equation holds, and here we are documenting the software. But in any case, I'd say just add the correct formula after a second equality in the equation in the Description if you want to keep it.

kahaaga replied:
> I don't think it matters for the software that this equation holds, and here we are documenting the software. But in any case, I'd say just add the correct formula after a second equality in the equation in the Description if you want to keep it.

I think it does matter for the software, because the ambiguous notation caused me to spend a lot of extra time during development and testing figuring out (also not realizing that was a potential issue) whether there was a typo in their definition, or if they made a computational mistake. 

> But in any case, I'd say just add the correct formula after a second equality in the equation in the Description if you want to keep it.

That's a good compromise.

Datseris reviewed (commented):
I've left several comments, so I'm stopping here!

The most important comment is about the generalized interface for normalizing entropies!

kahaaga commented on src/complexity_measures/reverse_dispersion_entropy.jl:
```diff
@@ -0,0 +1,91 @@
+end
+
+"""
+    reverse_dispersion(x::AbstractVector{T}, s = GaussianSymbolization(c = 5), m = 2, τ = 1,
+        normalize = true)
```
These should indeed be keyword arguments. I missed the semi-colon.
